### PR TITLE
MCR-2988 Add sourceUri validation in xEditor

### PR DESCRIPTION
**Pull request by Possommi**
Added property `MCR.XEditor.MCRXEditorTransformer.URISchema.disallowList=callJava`:

[Link to JIRA](https://mycore.atlassian.net/browse/MCR-2988).

### CODE REVIEW

File: mycore-xeditor/src/main/java/org/mycore/frontend/xeditor/MCRXEditorTransformer.java [hunk 1]
```diff
@@ -107,9 +114,25 @@ public void addNamespace(String prefix, String uri) {
...
+        if (!validateURI(uriRe)) {
+            throw new TransformerException("URI '" + uriRe + "' contains illegal uri scheme");
+        }
+
         editorSession.setEditedXML(uri);
     }
 
+    protected boolean validateURI(String uri) {
+        for (String scheme : MCRXEditorTransformer.DISALLOWED_URI_SCHEMAS) {
+            if (uri.contains(scheme)) {
```
Comment (rsteph-de) [inline mycore-xeditor/src/main/java/org/mycore/frontend/xeditor/MCRXEditorTransformer.java:128]: startsWith()?

File: mycore-xeditor/src/main/resources/components/xeditor/config/mycore.properties [hunk 2]
```diff
@@ -44,6 +44,9 @@ MCR.ContentTransformer.xeditor.Class=org.mycore.common.content.transformer.MCRXS
 MCR.ContentTransformer.xeditor.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl
 MCR.ContentTransformer.xeditor.Stylesheet=xsl/xeditor-templates.xsl,xsl/xeditor.xsl
 
+# Comma separated list of uri resolvers disallowed in the xed:source/@uri
+MCR.XEditor.MCRXEditorTransformer.URISchema.disallowList=callJava:
+
```
Comment (rsteph-de) [inline mycore-xeditor/src/main/resources/components/xeditor/config/mycore.properties:49]: with or without colon - I am not sure
If we keep the colon,  the value works like an URI prefix and we could define more complex URIs including server with server and pathes if needed
If we remove the colon, the value is a "proper" URI schema and we should at the colon in the validating Java code.

File: mycore-xeditor/src/main/java/org/mycore/frontend/xeditor/MCRXEditorTransformer.java (hunk 1 above)
Comment (Possommi) [reply mycore-xeditor/src/main/java/org/mycore/frontend/xeditor/MCRXEditorTransformer.java:128]: Does not catch xslStyle:xeditor-preprocessor:callJava:org.mycore.common.xml.MCRXMLFunctions:foo:bar

**Review by fluetze — CHANGES_REQUESTED** [review]
I will not approve this. It is not the responsibility of internal XEditor code to check URIs to be "allowed" or not. URIs are URIs, there can not be any assumption about their meanings or intended usage here. Setting allowed/disallowed patterns for the complete application will strictly prevent useful URIs even for admin use, for debugging etc. 

The misuse of URIs must be fixed by the application that uses the frameworks. There simply must be no line of code like xed:form/@uri="$foo" in the application. Instead, the application should only use parameters like in xed:form/@uri="mcrobject:{$id}" - and the mcrobject uri resolver will, hopefully, defend against nonsense object IDs.

**Comment by Possommi** [comment]
> I will not approve this. It is not the responsibility of internal XEditor code to check URIs to be "allowed" or not. URIs are URIs, there can not be any assumption about their meanings or intended usage here. Setting allowed/disallowed patterns for the complete application will strictly prevent useful URIs even for admin use, for debugging etc.
> 
> The misuse of URIs must be fixed by the application that uses the frameworks. There simply must be no line of code like xed:form/@uri="$foo" in the application. Instead, the application should only use parameters like in xed:form/@uri="mcrobject:{$id}" - and the mcrobject uri resolver will, hopefully, defend against nonsense object IDs.

It is true that one could check for a certain uri scheme and I could do that for applications under my control. But there are mycore based apps out there where their developers are not aware of this potential security issue. This fix would prevent at least the invocation of the callJava resolver.

**Comment by fluetze** [comment]
> It is true that one could check for a certain uri scheme and I could do that for applications under my control. But there are mycore based apps out there where their developers are not aware of this potential security issue. This fix would prevent at least the invocation of the callJava resolver.

The fix will only become effective it the application admin actually installs it. 
If he/she can do so, he/she could also fix the forms that contain the dangeroues line of code.
If mycore/mir/ubo use complete sourceURIs in provided code, that forms have to be fixed.

Do we also "fix" that in any other MyCoRe code that uses URIs? 
How do we know there isn't any application that calls that code?
What about custom URI resolvers? We should also check every XSL file.

**Comment by rsteph-de** [comment]
> Do we also "fix" that in any other MyCoRe code that uses URIs?
> How do we know there isn't any application that calls that code?
> What about custom URI resolvers? We should also check every XSL file.

We should at least try to identify the places where URIs are generated by user-input, e.g. HTML form input, request params, ...

**Comment by kkrebs** [comment]
is no xeditor problem, should not be fixed here, fix in application instead

**Pull request closed without merging**